### DISCUSSION TITLE
Inline buffer insertion/replacing possible without invoking prompt?

### DISCUSSION THREAD

**Discussion by ev-dev** [lzybkr/PSReadLine#3577] [Q&A]
Hey, does anyone know if/how it may be possible to insert text into the buffer without accepting/clearing it?

### More Details
Ideally, I want to insert some multi-line text into the prompt but still allow for further editing of that text.

Specifically, I'm using PowerShell v7 on Win11 in Windows Terminal and already have the ability to paste a multi-line blob of text into my prompt without invoking the command/text. 

So for instance, if I copy/paste this code into my prompt:
```powershell
0..9 |%{
  Write-Host "i=$_"

  # No closing bracket here
```
I am still able to use my arrow keys to move around & edit the text before invoking it via enter key (once all statements are terminated, otherwise a blank line is inserted).

I'm trying to recreate that same behavior but obviously not using the clipboard and instead via a PowerShell function I'm writing that would call the static methods from this lib. 

### What I've Tried
I've gone thru the examples in this repo (which are really cool btw), and I thought it would be as simple as using the `[Microsoft.PowerShell.PSConsoleReadLine]::Insert("0..9 |%{ $_")` or `::Replace(string)` methods, but alas, I'm struggling to keep the buffer open at all. 

Once I insert text with either of those methods, I can't seem to prevent the prompt from immediately clearing the buffer.

Anyone have any suggestions? Thanks in advance!

> PS. This is a small part of a PowerShell module I'm working on that seeks to simulate some of the behavior of the [fish shell](https://fishshell.com). Specifically, its [`funced` utility](https://fishshell.com/docs/current/cmds/funced.html) & built-in editor which allows you to very quickly create/edit user-defined functions in your prompt.

**Comment by Mattwmaster58** [comment]
Did you ever find an acceptable way to do this? Currently trying to implement [this gist for zsh](https://gist.github.com/iannuttall/d44907230f3c909173db7d15f8dccdc2) in powershell and running into a similar limitation

**Reply by Mattwmaster58** [reply]
So this is how I've solved it for my use case: insert the thing into the history, then use wshell to send the keys to bring up this history entry. The input gets buffered while it's own wshell command are running so when it finishes it brings up your history item very quickly.

```ps1
    [Microsoft.PowerShell.PSConsoleReadLine]::AddToHistory( $cmd.Trim())
    $wshell = New-Object -ComObject wscript.shell
    $wshell.SendKeys('{UP}')
    $wshell.SendKeys('{END}')
    $wshell.SendKeys('{LEFT}')
```

**Comment by sharpchen** [comment]
> Once I insert text with either of those methods, I can't seem to prevent the prompt from immediately clearing the buffer.

It sounds like you invoked the method in prompt directly? Of course it didn't work, you invoked it by hitting ENTER, the ENTER clears the previous buffer, not the method.
If you wrap it as a key binding, it will work.

```ps1
Set-PSReadLineKeyHandler -Chord 'ctrl+k' -ViMode Command -ScriptBlock {
    [Microsoft.PowerShell.PSConsoleReadLine]::Insert("0..9 |%{ $_")
}
```
